File: 백준/Gold/1300. K번째 수/K번째 수.java
import java.io.BufferedReader;
import java.io.InputStreamReader;

public class Main {
	static BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
	static int n,k;
	public static void main(String[] args) throws Exception {
		n = Integer.parseInt(br.readLine());
		k = Integer.parseInt(br.readLine());
		long start = 1;
		long end = k;
		while (start<=end) {
			long mid = start+(end-start)/2;
			long cnt=0;
			for (int i=1; i<=n; i++) {
				cnt+= Math.min(mid/i, n);
			}
			if (cnt < k){
				start = mid+1;
			}
			else {
				end = mid-1;
			}
		}
		System.out.println(start);
	}

}

File: 백준/Gold/1300. K번째 수/K번째 수.py
n = int(input())
k = int(input())
start = 1
end = k # k번째 수는 k 이하
while start <= end:
    mid = (start + end) // 2
    cnt = 0
    for i in range(1,n+1):
        cnt += min(n, mid//i) # 각 행에 있는 원소 개수 제한(n)
    if cnt >= k:
        end = mid - 1
    else:
        start = mid + 1
print(start)